File: backend/src/services/weaknessService.js
const supabase = require('../config/supabase')

/**
 * Analyze a user's message history to find their weakest topics per subject.
 * Returns: { physics: [{topic, error_rate, attempts}], chemistry: [...], mathematics: [...] }
 */
async function getWeaknessReport(userId, subject = null) {
  let query = supabase
    .from('messages')
    .select('topic_tags, is_correct, created_at')
    .eq('user_id', userId)
    .eq('role', 'user')
    .not('topic_tags', 'is', null)

  const { data: messages, error } = await query
  if (error) throw error

  // Combine with conversation subject info
  const { data: convData, error: convError } = await supabase
    .from('messages')
    .select('topic_tags, is_correct, conversations!inner(subject)')
    .eq('user_id', userId)
    .eq('role', 'assistant')
    .not('topic_tags', 'is', null)
    .order('created_at', { ascending: false })
    .limit(200)

  if (convError) throw convError

  // Aggregate: { [subject]: { [topic]: { attempts, errors } } }
  const stats = { physics: {}, chemistry: {}, mathematics: {} }

  for (const msg of convData || []) {
    const subj = msg.conversations?.subject || 'general'
    if (!stats[subj]) continue

    for (const tag of msg.topic_tags || []) {
      if (!stats[subj][tag]) stats[subj][tag] = { attempts: 0, errors: 0 }
      stats[subj][tag].attempts++
      if (msg.is_correct === false) stats[subj][tag].errors++
    }
  }

  // Convert to sorted array of weak topics (error_rate >= 0.3 and attempts >= 2)
  const result = {}
  for (const [subj, topics] of Object.entries(stats)) {
    if (subject && subj !== subject) continue
    result[subj] = Object.entries(topics)
      .filter(([_, v]) => v.attempts >= 2)
      .map(([topic, v]) => ({
        topic,
        attempts: v.attempts,
        error_rate: v.attempts > 0 ? v.errors / v.attempts : 0,
      }))
      .sort((a, b) => b.error_rate - a.error_rate)
  }

  return result
}

/**
 * Get the top N weakest topic names for a subject (to pass to quiz generator)
 */
async function getTopWeakTopics(userId, subject, n = 3) {
  const report = await getWeaknessReport(userId, subject)
  const topics = report[subject] || []

  if (topics.length === 0) {
    // Fallback: pick general topics for the subject
    const fallbacks = {
      physics: ['kinematics', 'thermodynamics', 'electrostatics'],
      chemistry: ['chemical-equilibrium', 'coordination-compounds', 'electrochemistry'],
      mathematics: ['integration', 'probability', 'matrices'],
    }
    return fallbacks[subject] || []
  }

  return topics.slice(0, n).map((t) => t.topic)
}

module.exports = { getWeaknessReport, getTopWeakTopics }

File: backend/app/services/weakness_service.py
from app.core.supabase_client import supabase

def get_weakness_report(user_id: str, subject: str = None) -> dict:
    query = supabase.table('messages') \
        .select('topic_tags, is_correct, conversations!inner(subject)') \
        .eq('user_id', user_id) \
        .eq('role', 'assistant') \
        .not_.is_('topic_tags', 'null') \
        .order('created_at', desc=True) \
        .limit(200) \
        .execute()
    
    conv_data = query.data or []
    stats = {'physics': {}, 'chemistry': {}, 'mathematics': {}, 'general': {}}
    
    for msg in conv_data:
        subj = msg.get('conversations', {}).get('subject') or 'general'
        if subj not in stats: continue
        tags = msg.get('topic_tags') or []
        for tag in tags:
            if tag not in stats[subj]:
                stats[subj][tag] = {'attempts': 0, 'errors': 0}
            stats[subj][tag]['attempts'] += 1
            if msg.get('is_correct') is False:
                stats[subj][tag]['errors'] += 1

    result = {}
    for subj_key, topics in stats.items():
        if subject and subj_key != subject: continue
        topic_list = []
        for topic, v in topics.items():
            topic_list.append({
                'topic': topic,
                'attempts': v['attempts'],
                'error_rate': v['errors'] / v['attempts'] if v['attempts'] > 0 else 0
            })
        topic_list.sort(key=lambda x: x['error_rate'], reverse=True)
        result[subj_key] = topic_list
    return result

def get_top_weak_topics(user_id: str, subject: str, n: int = 3) -> list[str]:
    report = get_weakness_report(user_id, subject)
    topics = report.get(subject, [])
    if not topics:
        fallbacks = {
            'physics': ['kinematics', 'thermodynamics'],
            'chemistry': ['equilibrium', 'bonding'],
            'mathematics': ['calculus', 'algebra'],
            'general': ['basics']
        }
        return fallbacks.get(subject, fallbacks['general'])
    return [t['topic'] for t in topics[:n]]
